test/jdk/java/util/concurrent/tck/ForkJoinPool20Test.java

```diff
@@ -216,4 +238,416 @@ static ForkJoinWorkerThread submitBusyTask(ForkJoinPool pool) throws Exception {
+        Thread.sleep(timeoutMillis());
+        assertTrue(task.isCancelled());
+    }
+
+    static final class SubmitWithTimeoutException extends RuntimeException {}
+
+    /**
+     * submitWithTimeout using complete completes after timeout
+     */
+    public void testSubmitWithCompleterTimeoutCompletes() throws InterruptedException {
+        final ForkJoinPool p = ForkJoinPool.commonPool();
+        Callable<Item> c = new Callable<Item>() {
+            public Item call() throws Exception {
+                Thread.sleep(LONGER_DELAY_MS); return one; }};
+        ForkJoinTask<Item> task = p.submitWithTimeout(
+            c, 1, NANOSECONDS,
+            (ForkJoinTask<Item> t) ->
+            t.complete(two));
+        Thread.sleep(timeoutMillis());
+        assertEquals(task.join(), two);
+    }
+
+    /**
+     * submitWithTimeout using completeExceptionally throws after timeout
+     */
+    public void testSubmitWithTimeoutThrows() throws InterruptedException {
+        final ForkJoinPool p = ForkJoinPool.commonPool();
+        Callable<Boolean> c = new Callable<Boolean>() {
+            public Boolean call() throws Exception {
+                Thread.sleep(LONGER_DELAY_MS); return Boolean.TRUE; }};
+        ForkJoinTask<Boolean> task = p.submitWithTimeout(
+            c, 1, NANOSECONDS,
+            (ForkJoinTask<Boolean> t) ->
+            t.completeExceptionally(new SubmitWithTimeoutException()));
+        Thread.sleep(timeoutMillis());
+        try {
+            task.join();
+            shouldThrow();
+        }
+        catch (Exception ex) {
+            assertTrue(ex instanceof SubmitWithTimeoutException);
+        }
+    }
+
+    /**
+     * submitWithTimeout doesn't cancel if completed before timeout
+     */
+    public void testSubmitWithTimeout_NoTimeout() throws InterruptedException {
+        final ForkJoinPool p = ForkJoinPool.commonPool();
+        Callable<Boolean> c = new Callable<Boolean>() {
+            public Boolean call() throws Exception {
+                return Boolean.TRUE; }};
+        ForkJoinTask<?> task = p.submitWithTimeout(c, LONGER_DELAY_MS, MILLISECONDS, null);
+        Thread.sleep(timeoutMillis());
+        assertFalse(task.isCancelled());
+    }
+
+    /**
+     * A delayed task completes (possibly abnormally) if shutdown after
+     * calling cancelDelayedTasksOnShutdown()
+     */
+    public void testCancelDelayedTasksOnShutdown() throws Exception {
+        final ForkJoinPool p = new ForkJoinPool(2);
+        p.cancelDelayedTasksOnShutdown();
+        try (PoolCleaner cleaner = cleaner(p)) {
+            Callable<Boolean> task = new CheckedCallable<>() {
+                public Boolean realCall() {
+                    return Boolean.TRUE;
+                }};
+            Future<Boolean> f = p.schedule(task, LONGER_DELAY_MS, MILLISECONDS);
+            p.shutdown();
+            assertTrue(p.awaitTermination(LONG_DELAY_MS, MILLISECONDS));
+            assertTrue(f.isDone());
```

Comment:
Might make sense to factor out assertions of sets of states associated with successful, cancelled, failed and use those asserts throughout these tests (just to make sure that there aren't weird combinations of states).

Reply:
Right. We have some helper methods along these lines, but in the wrong test files (mainly ForkJoinTaskTest), so would need some refactoring to use here. Someday.